marathon/src/main/scala/io/buoyant/marathon/v2/Api.scala

```diff
@@ -25,46 +28,75 @@ object Api {
 
   val versionString = "v2"
 
-  private[this] case class SetHost(host: String)
-    extends SimpleFilter[http.Request, http.Response] {
+  case class AuthRequest(
+    loginEndpoint: String,
+    uid: String,
+    privateKey: String
+  ) {
+    private[this] val jwt = Jwt.encode(s"""{"uid":"$uid"}""", privateKey, JwtAlgorithm.RS256)
 
-    def apply(req: http.Request, service: Service[http.Request, http.Response]) = {
-      req.host = host
-      service(req)
-    }
+    val path = new URL(loginEndpoint).getPath
+    val jwtToken = s"""{"uid":"$uid","token":"$jwt"}"""
   }
 
-  def apply(client: Client, host: String, uriPrefix: String): Api =
-    new AppIdApi(SetHost(host).andThen(client), s"$uriPrefix/$versionString")
+  def apply(client: Client, host: String, uriPrefix: String, auth: Option[AuthRequest]): Api =
+    new AppIdApi(SetHost(host).andThen(client), s"$uriPrefix/$versionString", auth)
+
+  private[v2] val log = Logger.get(getClass.getName)
 
   private[v2] def rspToApps(rsp: http.Response): Future[Api.AppIds] =
     rsp.status match {
       case http.Status.Ok =>
         val apps = readJson[AppsRsp](rsp.content).map(_.toApps)
         Future.const(apps)
-
-      case _ => Future.exception(UnexpectedResponse(rsp))
+      case http.Status.Unauthorized =>
+        val e = UnauthorizedResponse(rsp)
+        log.error(s"rspToApps returned Unauthorized with $e")
+        Future.exception(e)
+      case _ =>
+        val e = UnexpectedResponse(rsp)
+        log.error(s"rspToApps returned Unexpected with $e")
+        Future.exception(e)
     }
 
   private[v2] def rspToAddrs(rsp: http.Response): Future[Set[Address]] =
     rsp.status match {
       case http.Status.Ok =>
         val addrs = readJson[AppRsp](rsp.content).map(_.toAddresses)
         Future.const(addrs)
+      case http.Status.Unauthorized =>
+        val e = UnauthorizedResponse(rsp)
+        log.error(s"rspToAddrs returned Unauthorized with $e")
+        Future.exception(e)
       case _ =>
-        Future.exception(UnexpectedResponse(rsp))
+        val e = UnexpectedResponse(rsp)
+        log.error(s"rspToAddrs returned Unexpected with $e")
+        Future.exception(e)
```

Comment:
Let's remove these logs in favor of logging centrally once for the failure (i.e where it's called).

Note that it's preferred to use `log.error(e, ...` so that we properly formatted stack traces, etc.  Then, we don't need to have programmer-focused details like "rspToAddrs" in the log message, since that is captured in the stack trace.

Ideally, would like the log message to be something like _marathon at $host returned $code_